### PR TITLE
MAID-2761: add all structs related to gossiping messages

### PR DESCRIPTION
**Pull request by Fraser999**

### CODE REVIEW

File: src/block.rs [hunk 1]
```diff
@@ -50,7 +49,10 @@ impl<T: Serialize + DeserializeOwned + Debug + Eq, P: PublicId> Block<T, P> {
         &self.proofs
     }
 
-    /// Convert the vote to a proof and insert it into proofs of the block.
+    /// Converts `vote` to a `Proof` and attempts to add it to the block.  Returns an error if
+    /// `vote` is invalid (i.e. signature check fails or the `vote` is for a different network
+    /// event, `Ok(true)` if the `Proof` wasn't previously held in this `Block`, or `Ok(false)` if
```
Comment (maqi) [inline src/block.rs]: is there a missing `)` after the word of `event` ?

File: src/error.rs [hunk 2]
```diff
@@ -20,12 +20,6 @@ quick_error! {
             description("Signature cannot be verified")
             display("The message or signature might be corrupted, or the signer is wrong.")
         }
-        /// IO error.
-        Io(error: ::std::io::Error) {
-            description(error.description())
-            display("I/O error: {}", error)
-            from()
-        }
```
Comment (maqi) [inline src/error.rs]: the crate may need to dump the stuff to a file eventually. so maybe this IO error is still required?

File: src/gossip/event.rs [hunk 3]
```diff
@@ -6,23 +6,133 @@
...
+            Hash::from(serialised_content.as_slice())
+        } else {
+            return Err(Error::SignatureFailure);
+        };
+
+        Ok(Self {
+            content: packed_event.content,
+            signature: packed_event.signature,
+            hash,
+            index: Some(0),
```
Comment (maqi) [inline src/gossip/event.rs]: I assume this index needs to be updated outside?
shall we make a comment here to explain it?
and same for the following `new` function.

File: src/error.rs (hunk 2 above)
Comment (Fraser999) [reply src/error.rs]: I'd prefer to leave it out until we need it.

File: src/gossip/event.rs (hunk 3 above)
Comment (Fraser999) [reply src/gossip/event.rs]: Fair point.  I'll actually change the `index` to be initialised to `None` to make this clearer too.

File: src/peer_manager.rs [hunk 4]
```diff
@@ -86,17 +83,35 @@ impl<S: SecretId> PeerManager<S> {
...
     }
 
-    /// Add event created by the peer. Returns an error if `peer` is not known, `Ok(None)` if the
-    /// `index` wasn't previously held for that peer, or `Ok(Some(hash))` if it was previously held
-    /// and the previous hash is returned.
-    pub fn add_event<T: Serialize + DeserializeOwned + Debug + Eq, P: PublicId>(
+    /// Adds event created by the peer. Returns an error if `peer_id` is not known, or if we already
+    /// held an event from this peer with this index, but that event's hash is different to the one
+    /// being added (in which case `peers` is left unmodified).
+    pub fn add_event<T: NetworkEvent, P: PublicId>(
```
Comment (maqi) [inline src/peer_manager.rs]: We don't have `remove_event` function.
Which means the first coming event will occupy the position forever and cannot be corrected.
Will there be chance such correction will be required to carry out?

Comment (Fraser999) [reply src/peer_manager.rs]: I think the plan is that we should only be adding events here once we know they're valid.  For example, to add event 4, we'd need to have event 3 already so we can validate 4's `self_parent` is 3.  Or if we're about add an `Initial` event for a peer, we should check we don't already have one.  So really the `InvalidEventIndex` error shouldn't ever get returned.

**Pull request merged**